File: app/src/main/java/com/qiaoqiao/repository/knowledge/DsKnowledgeRemoteSource.java
package com.qiaoqiao.repository.knowledge;


import android.support.annotation.NonNull;

import com.google.android.gms.maps.model.LatLng;
import com.qiaoqiao.app.Key;
import com.qiaoqiao.repository.AbstractDsSource;
import com.qiaoqiao.repository.DsLoadedCallback;
import com.qiaoqiao.repository.annotation.RepositoryScope;
import com.qiaoqiao.repository.backend.Google;
import com.qiaoqiao.repository.backend.Wikipedia;
import com.qiaoqiao.repository.backend.model.wikipedia.LangLink;

import java.util.Locale;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@RepositoryScope
public final class DsKnowledgeRemoteSource extends AbstractDsSource {
	private @NonNull  final Key mKey;

	public DsKnowledgeRemoteSource(@NonNull Key key, @NonNull Google google, @NonNull Wikipedia wikipedia) {
		super(google, wikipedia);
		mKey = key;
	}

	@Override
	public void onTranslate(@NonNull String q, @NonNull DsLoadedCallback callback) {
		getGoogle().getTranslateService()
		           .translate(q,
		                      Locale.getDefault()
		                            .getLanguage(),
		                      "text",
		                      mKey.toString())
		           .subscribeOn(Schedulers.io())
		           .observeOn(AndroidSchedulers.mainThread())
		           .subscribe(response -> callback.onTranslateData(response.getData()));
	}


	@Override
	public void onKnowledgeQuery(@NonNull String keyword, @NonNull DsLoadedCallback callback) {
		getWikipedia().getResult(new Wikipedia.WikiReqBody(Locale.getDefault()
		                                                         .getLanguage(), keyword))
		              .subscribeOn(Schedulers.io())
		              .observeOn(AndroidSchedulers.mainThread())
		              .subscribe(result1 -> {
			              try {
				              if (result1.getQuery()
				                         .getPages()
				                         .getList()
				                         .size() > 0) {
					              callback.onKnowledgeResponse(result1);
				              }
			              } catch (Exception e) {
				              callback.onException(e);
			              }
		              });
	}

	@Override
	public void onKnowledgeQuery(@NonNull LangLink langLink, @NonNull DsLoadedCallback callback) {
		getWikipedia().getResult(new Wikipedia.WikiReqBody(langLink.getLanguage(), langLink.getQuery()))
		              .subscribeOn(Schedulers.io())
		              .observeOn(AndroidSchedulers.mainThread())
		              .subscribe(result1 -> {
			              try {
				              if (result1.getQuery()
				                         .getPages()
				                         .getList()
				                         .size() > 0) {
					              callback.onKnowledgeResponse(result1);
				              }
			              } catch (Exception e) {
				              callback.onException(e);
			              }
		              });
	}

	@Override
	public void onGeosearchQuery(@NonNull LatLng latLng, @NonNull DsLoadedCallback callback) {
		String geoLoc = String.format("%s|%s", latLng.latitude + "", latLng.longitude + "");
		getWikipedia().getGeosearch(wikiGeosearch(Locale.getDefault()
		                                                            .getLanguage(), geoLoc))
		              .subscribeOn(Schedulers.io())
		              .observeOn(AndroidSchedulers.mainThread())
		              .subscribe(result1 -> {
			              try {
				              if (result1.getQuery() != null && result1.getQuery()
				                                                       .getGeosearches().length > 0) {
					              callback.onGeosearchResponse(result1);
				              }
			              } catch (Exception e) {
				              callback.onException(e);
			              }
		              });
	}

	private @NonNull
	String wikiQuery(@NonNull String lang, @NonNull String keyword) {
		return wikiHost(lang) + wikiQuery(keyword);
	}

	private @NonNull
	String wikiHost(@NonNull String lang) {
		return String.format("https://%s.wikipedia.org", lang);
	}

	private @NonNull
	String wikiQuery(@NonNull String keyword) {
		return "/w/api.php?format=json&action=query&prop=extracts|pageimages|langlinks&llprop=autonym&lldir=descending&lllimit=500&piprop=original|name|thumbnail&exlimit=1&redirects=titles&pageids="
				+ keyword;
	}

	private @NonNull
	String wikiGeosearch(@NonNull String lang, @NonNull String keyword) {
		return wikiHost(lang) + wikiGeosearch(keyword);
	}


	private @NonNull
	String wikiGeosearch(@NonNull String keyword) {
		return "/w/api.php?format=json&action=query&list=geosearch&gsradius=10000&gslimit=max&gscoord=" + keyword;
	}

}

File: app/src/main/java/com/qiaoqiao/repository/knowledge/DsKnowledgeRemoteSource.kt
package com.qiaoqiao.repository.knowledge

import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.vision.barcode.Barcode
import com.qiaoqiao.app.Key
import com.qiaoqiao.core.product.model.ProductEntity
import com.qiaoqiao.repository.AbstractDsSource
import com.qiaoqiao.repository.DsLoadedCallback
import com.qiaoqiao.repository.backend.Google
import com.qiaoqiao.repository.backend.KnowledgeRequest
import com.qiaoqiao.repository.backend.ProductsService
import com.qiaoqiao.repository.backend.Wikipedia
import com.qiaoqiao.repository.backend.model.wikipedia.LangLink
import com.qiaoqiao.rx.IoToMainScheduleObservable
import io.reactivex.disposables.Disposable
import java.util.*

class DsKnowledgeRemoteSource(private val key: Key, google: Google, wikipedia: Wikipedia, productsService: ProductsService) : AbstractDsSource(google, wikipedia, productsService) {
    override fun onTranslate(q: String, callback: DsLoadedCallback): Disposable {
        google?.let {
            return (it.translateService
                    .translate(q,
                            Locale.getDefault().language,
                            "text",
                            key.toString())
                    .compose(IoToMainScheduleObservable())
                    .subscribe({ callback.onTranslateData(it.data) }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }

    override fun onKnowledgeQuery(keyword: String, callback: DsLoadedCallback): Disposable {
        wikipedia?.let {
            return (it.getResult(KnowledgeRequest(Locale.getDefault()
                    .language, keyword))
                    .compose(IoToMainScheduleObservable())
                    .subscribe({
                        if (it.query
                                .pages
                                .list
                                .size > 0) {
                            callback.onKnowledgeResponse(it)
                        }
                    }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }

    override fun onKnowledgeQuery(langLink: LangLink, callback: DsLoadedCallback): Disposable {
        wikipedia?.let {
            return (it.getResult(KnowledgeRequest(langLink.language, langLink.query))
                    .compose(IoToMainScheduleObservable())
                    .subscribe({
                        if (it.query
                                .pages
                                .list
                                .size > 0) {
                            callback.onKnowledgeResponse(it)
                        }
                    }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }

    override fun onKnowledgeQuery(pageId: Int, callback: DsLoadedCallback): Disposable {
        wikipedia?.let {
            return (it.getResult(wikiQuery(Locale.getDefault()
                    .language, pageId.toString() + ""))
                    .compose(IoToMainScheduleObservable())
                    .subscribe({
                        if (!it.query.pages.list.isEmpty()) {
                            callback.onKnowledgeResponse(it)
                        }
                    }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }

    override fun onGeosearchQuery(latLng: LatLng, radius: Long, callback: DsLoadedCallback): Disposable {
        wikipedia?.let {
            val geoLoc = String.format("%s|%s", latLng.latitude.toString(), latLng.longitude.toString())
            return (it.getGeosearch(wikiGeosearch(Locale.getDefault()
                    .language, radius, geoLoc))
                    .compose(IoToMainScheduleObservable())
                    .subscribe({
                        if (it.query != null) {
                            callback.onGeosearchResponse(it)
                        }
                    }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }

    override fun onKnowledgeQuery(barcode: Barcode, callback: DsLoadedCallback): Disposable {
        productsService?.let {
            return (it.getProduct(KnowledgeRequest(Locale.getDefault().language, barcode.rawValue)).compose(IoToMainScheduleObservable())
                    .subscribe({
                        callback.onKnowledgeResponse(ProductEntity(it))
                    }, { callback.onException(it) }))
        } ?: run {
            throw NotImplementedError()
        }
    }
}

private fun wikiQuery(lang: String, keyword: String) = wikiHost(lang) + wikiQuery(keyword)
private fun wikiHost(lang: String) = String.format("https://%s.wikipedia.org", lang)
private fun wikiQuery(keyword: String) = "/w/api.php?format=json&action=query&prop=extracts|pageimages|langlinks&llprop=autonym&lldir=descending&lllimit=500&piprop=original|name|thumbnail&exlimit=1&redirects=titles&pageids=" + keyword
private fun wikiGeosearch(lang: String, radius: Long, keyword: String) = wikiHost(lang) + wikiGeosearch(radius, keyword)
private fun wikiGeosearch(radius: Long, keyword: String) = "/w/api.php?format=json&action=query&list=geosearch&gsradius=$radius&gslimit=max&gscoord=$keyword"
